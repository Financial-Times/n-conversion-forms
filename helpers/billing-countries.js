import { billingCountries } from 'n-common-static-data';

export { billingCountries };
