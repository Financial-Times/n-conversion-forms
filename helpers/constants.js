const printRegions = {
	cemeaV1: 'CEMEA_V1',
	cemeaV2: 'CEMEA_V2',
	apac: 'APAC',
};

const addressLine1AutoComplete = 'address-line1';
const addressLine2AutoComplete = 'address-line2';
const addressLine3AutoComplete = 'address-line3';

module.exports = {
	printRegions,
	addressLine1AutoComplete,
	addressLine2AutoComplete,
	addressLine3AutoComplete,
};
