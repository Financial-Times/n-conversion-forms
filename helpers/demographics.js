const { demographics } = require('n-common-static-data');

export { demographics };
