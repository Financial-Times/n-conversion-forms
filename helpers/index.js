module.exports = {
	'ncf-countries': require('./ncf-countries'),
	'ncf-common-data': require('./ncf-common-data'),
	cemeaV1ISO: require('./supportedCountries').cemeaV1ISO,
	cemeaV2ISO: require('./supportedCountries').cemeaV2ISO,
	apacISO: require('./supportedCountries').apacISO,
	identifyFTShippingZone: require('./supportedCountries')
		.identifyFTShippingZone,
	countriesSupported: require('./supportedCountries').countriesSupported,
	countriesSupportedISO: require('./supportedCountries').countriesSupportedISO,
	supportedPostcodeValidators: require('./supportedPostcodeValidators')
		.supportedPostcodeValidators,
	allSupportedPostcodeExamples: require('./supportedPostcodeExamples')
		.allSupportedPostcodeExamples,
	demographics: require('./demographics').demographics,
	billingCountries: require('./billing-countries').billingCountries,
};
