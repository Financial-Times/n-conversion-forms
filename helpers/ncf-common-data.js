const _get = require('lodash.get');
const COMMON_DATA = require('n-common-static-data');

module.exports = function ({ hash, fn }) {
	const context = Object.assign({}, { [hash.export]: _get(COMMON_DATA, hash.import) }, this);
	return fn(context);
};
