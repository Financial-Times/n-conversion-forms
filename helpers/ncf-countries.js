const { countries } = require('n-common-static-data').billingCountries;

const MINIMUM_TO_GROUP = 50;
const MINIMUM_TO_SHOW_GROUPS = 2;

module.exports = function ({ hash = {}, fn }) {
	const filter = Array.isArray(hash.filterList) ? hash.filterList : false;
	const value = hash.value;

	let data = countries;

	// Only show countries in the filter
	if (filter) {
		data = data.filter(item => filter.includes(item.code));
	}

	// Select the country
	if (value) {
		data = data.map(item => {
			item.selected = item.code === value;
			return item;
		});
	}

	// Group countries
	if (data.length >= MINIMUM_TO_GROUP) {
		data = groupCountries(data);
	}

	const context = Object.assign({ countries: data }, this);
	return fn(context);
};

/**
 * Produce a frequently used option group
 * @param {Array} countries
 * @return {Array}
 */
function groupCountries (countries) {
	const frequentlyUsed = ['GBR', 'USA', 'JPN', 'FRA', 'CAN'];

	const frequentlyUsedCountries = countries
		.filter(item => frequentlyUsed.includes(item.code))
		.map(item => Object.assign({}, item))
		.sort((a, b) => frequentlyUsed.indexOf(a.code) - frequentlyUsed.indexOf(b.code));

	// Not worth showing return standard countries list
	if (frequentlyUsedCountries.length < MINIMUM_TO_SHOW_GROUPS) {
		return countries;
	}

	// If selected country is in frequently used only leave that one selected
	// otherwise the user is shown the lowest selected option
	const selected = frequentlyUsedCountries.find(item => item.selected);
	if (selected) {
		countries.forEach(item => item.selected = false);
	}

	return [
		{ label: 'Frequently Used', countries: frequentlyUsedCountries },
		{ label: 'Alphabetical', countries }
	];
}
