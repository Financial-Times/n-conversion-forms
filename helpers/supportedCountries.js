const supportedCountriesMasterList = {
	cemeaV1: {
		AUT: 'Austria',
		BEL: 'Belgium',
		BGR: 'Bulgaria',
		HRV: 'Croatia',
		CZE: 'Czechia',
		DNK: 'Denmark',
		FIN: 'Finland',
		DEU: 'Germany',
		GRC: 'Greece',
		HUN: 'Hungary',
		ISL: 'Iceland',
		ITA: 'Italy',
		LIE: 'Liechtenstein',
		LTU: 'Lithuania',
		NLD: 'Netherlands',
		NOR: 'Norway',
		POL: 'Poland',
		PRT: 'Portugal',
		QAT: 'Qatar',
		ROU: 'Romania',
		RUS: 'Russian Federation',
		SVK: 'Slovakia',
		SVN: 'Slovenia',
		ZAF: 'South Africa',
		ESP: 'Spain',
		SWE: 'Sweden',
		CHE: 'Switzerland'
	},
	cemeaV2: { FRA: 'France', LUX: 'Luxembourg', MCO: 'Monaco' },
	apac: {
		CHN: 'China',
		HKG: 'Hong Kong',
		IND: 'India',
		IDN: 'Indonesia',
		JPN: 'Japan',
		KOR: 'Korea, Republic of',
		MYS: 'Malaysia',
		PHL: 'Philippines',
		SGP: 'Singapore',
		TWN: 'Taiwan, Province of China',
		THA: 'Thailand'
	},
	GBR: 'United Kingdom',
	USA: 'United States',
	CAN: 'Canada'
};


const cemeaV1ISO = Object.keys(supportedCountriesMasterList.cemeaV1);
const cemeaV2ISO = Object.keys(supportedCountriesMasterList.cemeaV2);
const apacISO = Object.keys(supportedCountriesMasterList.apac);
const countriesSupported = Object.assign({}, ...(function _flatten (o) { return [].concat(...Object.keys(o).map(k => typeof o[k] === 'object' ? _flatten(o[k]) : ({[k]: o[k]})));}(supportedCountriesMasterList)));


module.exports = { cemeaV1ISO, cemeaV2ISO, apacISO, countriesSupported };
