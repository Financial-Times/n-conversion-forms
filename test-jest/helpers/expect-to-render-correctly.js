import ReactDOMServer from 'react-dom/server';

export const expectToRenderCorrectly = {
	toRenderCorrectly(reactComponent, props) {
		// Render the React component with props.
		const ReactComponentRendered = ReactDOMServer.renderToStaticMarkup(
			reactComponent(props)
		);

		// eslint-disable-next-line jest/no-standalone-expect
		expect(ReactComponentRendered).toMatchSnapshot();

		// This point is reached when the above assertions were successful.
		// The test should therefore always pass, that means it needs to be
		// `true` when used normally, and `false` when `.not` was used.
		return {
			pass: !this.isNot,
		};
	},
};
