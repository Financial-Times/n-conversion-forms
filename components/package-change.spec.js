import { PackageChange } from './index';
import { expectToRenderCorrectly } from '../test-jest/helpers/expect-to-render-correctly';

expect.extend(expectToRenderCorrectly);

describe('PackageChange', () => {
	it('render with defaults', () => {
		const props = {
			changePackageUrl: 'https://www.ft.com',
			currentPackage: 'Trial',
		};

		expect(PackageChange).toRenderCorrectly(props);
	});

	it('renders with description', () => {
		const props = {
			changePackageUrl: 'https://www.ft.com',
			currentPackage: 'Trial',
			packageDescription: 'Personalised email briefings and alerts',
		};

		expect(PackageChange).toRenderCorrectly(props);
	});

	['annual'].forEach((term) => {
		describe(`${term}`, () => {
			it('render with defaults', () => {
				const props = {
					changePackageUrl: 'https://www.ft.com',
					currentPackage: 'Trial',
					terms: [
						{
							name: term,
							price: '£1.00',
							weeklyPrice: '£1.00',
						},
					],
				};

				expect(PackageChange).toRenderCorrectly(props);
			});

			it('render with discount', () => {
				const props = {
					changePackageUrl: 'https://www.ft.com',
					currentPackage: 'Trial',
					terms: [
						{
							name: term,
							price: '£1.00',
							weeklyPrice: '£1.00',
							discount: '25%',
						},
					],
				};

				expect(PackageChange).toRenderCorrectly(props);
			});

			it('render with trialPrice', () => {
				const props = {
					changePackageUrl: 'https://www.ft.com',
					currentPackage: 'Trial',
					terms: [
						{
							name: term,
							price: '£1.00',
							weeklyPrice: '£1.00',
							trialPrice: '£1.00',
						},
					],
				};

				expect(PackageChange).toRenderCorrectly(props);
			});

			it('render when is7DayPassExperiment is true', () => {
				const props = {
					changePackageUrl: 'https://www.ft.com',
					currentPackage: 'Trial',
					is7DayPassExperiment: true,
					terms: [
						{
							name: term,
							price: '£4.99',
							weeklyPrice: '£4.99',
						},
					],
				};

				expect(PackageChange).toRenderCorrectly(props);
			});
		});
	});
});
