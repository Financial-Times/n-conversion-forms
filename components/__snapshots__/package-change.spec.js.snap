// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PackageChange annual render with defaults 1`] = `
<div class="ncf__package-change">
  <div class="ncf__package-change__package">
    <p class="ncf__package-change__content">
      You have chosen
      <span class="ncf__strong">
        Trial
      </span>
    </p>
    <div class="ncf__package-change__actions">
      <a href="https://www.ft.com"
         class="ncf__button ncf__button--mono ncf__button--baseline"
         data-trackable="change"
      >
        Change
      </a>
    </div>
  </div>
  <div class="ncf__package-change__terms">
    <div class="ncf__package-change__term">
      <span class="ncf__package-change__title">
        Annually
      </span>
      <div class="ncf__package-change__description">
        Single
        <span class="ncf__package-change__price">
          £1.00
        </span>
        payment*
      </div>
      <div class="ncf__package-change__weekly-price">
        Just
        <span class="ncf__package-change__price">
          £1.00
        </span>
        per week
      </div>
    </div>
    <div class="ncf__package-change__annual-copy">
      * Save up to 25% when you pay annually. Discount varies across member states and across our subscription packages.
    </div>
  </div>
</div>
`;

exports[`PackageChange annual render with defaults 2`] = `
<div class="ncf__package-change">
  <div class="ncf__package-change__package">
    <p class="ncf__package-change__content">
      You have chosen
      <span class="ncf__strong">
        Trial
      </span>
    </p>
    <div class="ncf__package-change__actions">
      <a href="https://www.ft.com"
         class="ncf__button ncf__button--mono ncf__button--baseline"
         data-trackable="change"
      >
        Change
      </a>
    </div>
  </div>
  <div class="ncf__package-change__terms">
    <div class="ncf__package-change__term">
      <span class="ncf__package-change__title">
        Annually
      </span>
      <div class="ncf__package-change__description">
        Single
        <span class="ncf__package-change__price">
          £1.00
        </span>
        payment*
      </div>
      <div class="ncf__package-change__weekly-price">
        Just
        <span class="ncf__package-change__price">
          £1.00
        </span>
        per week
      </div>
    </div>
    <div class="ncf__package-change__annual-copy">
      * Save up to 25% when you pay annually. Discount varies across member states and across our subscription packages.
    </div>
  </div>
</div>
`;

exports[`PackageChange annual render with discount 1`] = `
<div class="ncf__package-change">
  <div class="ncf__package-change__package">
    <p class="ncf__package-change__content">
      You have chosen
      <span class="ncf__strong">
        Trial
      </span>
    </p>
    <div class="ncf__package-change__actions">
      <a href="https://www.ft.com"
         class="ncf__button ncf__button--mono ncf__button--baseline"
         data-trackable="change"
      >
        Change
      </a>
    </div>
  </div>
  <div class="ncf__package-change__terms">
    <div class="ncf__package-change__term ncf__package-change__term--discounted">
      <span class="ncf__package-change__title">
        Annually
      </span>
      <span class="ncf__package-change__discount">
        Save 25%
      </span>
      <div class="ncf__package-change__description">
        Single
        <span class="ncf__package-change__price">
          £1.00
        </span>
        payment*
      </div>
      <div class="ncf__package-change__weekly-price">
        Just
        <span class="ncf__package-change__price">
          £1.00
        </span>
        per week
      </div>
    </div>
  </div>
</div>
`;

exports[`PackageChange annual render with discount 2`] = `
<div class="ncf__package-change">
  <div class="ncf__package-change__package">
    <p class="ncf__package-change__content">
      You have chosen
      <span class="ncf__strong">
        Trial
      </span>
    </p>
    <div class="ncf__package-change__actions">
      <a href="https://www.ft.com"
         class="ncf__button ncf__button--mono ncf__button--baseline"
         data-trackable="change"
      >
        Change
      </a>
    </div>
  </div>
  <div class="ncf__package-change__terms">
    <div class="ncf__package-change__term ncf__package-change__term--discounted">
      <span class="ncf__package-change__title">
        Annually
      </span>
      <span class="ncf__package-change__discount">
        Save 25%
      </span>
      <div class="ncf__package-change__description">
        Single
        <span class="ncf__package-change__price">
          £1.00
        </span>
        payment*
      </div>
      <div class="ncf__package-change__weekly-price">
        Just
        <span class="ncf__package-change__price">
          £1.00
        </span>
        per week
      </div>
    </div>
  </div>
</div>
`;

exports[`PackageChange annual render with trialPrice 1`] = `
<div class="ncf__package-change">
  <div class="ncf__package-change__package">
    <p class="ncf__package-change__content">
      You have chosen
      <span class="ncf__strong">
        Trial
      </span>
    </p>
    <div class="ncf__package-change__actions">
      <a href="https://www.ft.com"
         class="ncf__button ncf__button--mono ncf__button--baseline"
         data-trackable="change"
      >
        Change
      </a>
    </div>
  </div>
  <div class="ncf__package-change__terms">
    <div class="ncf__package-change__term">
      <span class="ncf__package-change__title">
        Annually
      </span>
      <div class="ncf__package-change__description">
        Single
        <span class="ncf__package-change__price">
          £1.00
        </span>
        payment*
      </div>
      <div class="ncf__package-change__weekly-price">
        Just
        <span class="ncf__package-change__price">
          £1.00
        </span>
        per week
      </div>
    </div>
    <div class="ncf__package-change__annual-copy">
      * Save up to 25% when you pay annually. Discount varies across member states and across our subscription packages.
    </div>
  </div>
</div>
`;

exports[`PackageChange annual render with trialPrice 2`] = `
<div class="ncf__package-change">
  <div class="ncf__package-change__package">
    <p class="ncf__package-change__content">
      You have chosen
      <span class="ncf__strong">
        Trial
      </span>
    </p>
    <div class="ncf__package-change__actions">
      <a href="https://www.ft.com"
         class="ncf__button ncf__button--mono ncf__button--baseline"
         data-trackable="change"
      >
        Change
      </a>
    </div>
  </div>
  <div class="ncf__package-change__terms">
    <div class="ncf__package-change__term">
      <span class="ncf__package-change__title">
        Annually
      </span>
      <div class="ncf__package-change__description">
        Single
        <span class="ncf__package-change__price">
          £1.00
        </span>
        payment*
      </div>
      <div class="ncf__package-change__weekly-price">
        Just
        <span class="ncf__package-change__price">
          £1.00
        </span>
        per week
      </div>
    </div>
    <div class="ncf__package-change__annual-copy">
      * Save up to 25% when you pay annually. Discount varies across member states and across our subscription packages.
    </div>
  </div>
</div>
`;

exports[`PackageChange render with defaults 1`] = `
<div class="ncf__package-change">
  <div class="ncf__package-change__package">
    <p class="ncf__package-change__content">
      You have chosen
      <span class="ncf__strong">
        Trial
      </span>
    </p>
    <div class="ncf__package-change__actions">
      <a href="https://www.ft.com"
         class="ncf__button ncf__button--mono ncf__button--baseline"
         data-trackable="change"
      >
        Change
      </a>
    </div>
  </div>
</div>
`;

exports[`PackageChange render with defaults 2`] = `
<div class="ncf__package-change">
  <div class="ncf__package-change__package">
    <p class="ncf__package-change__content">
      You have chosen
      <span class="ncf__strong">
        Trial
      </span>
    </p>
    <div class="ncf__package-change__actions">
      <a href="https://www.ft.com"
         class="ncf__button ncf__button--mono ncf__button--baseline"
         data-trackable="change"
      >
        Change
      </a>
    </div>
  </div>
</div>
`;
